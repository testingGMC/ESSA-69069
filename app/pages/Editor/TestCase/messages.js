/*
 * TestCase Messages
 *
 * This contains all the text for the TestCase component.
 */

import { defineMessages } from 'react-intl'

export const scope = 'app.components.TestCase'

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the TestCase component!',
  },
  test: {
    id: `${scope}.test`,
    defaultMessage: 'Test',
  },
  correctAnswer: {
    id: `${scope}.correctAnswer`,
    defaultMessage: 'Correct Answer',
  },
  wrongAnswer: {
    id: `${scope}.wrongAnswer`,
    defaultMessage: 'Wrong Answer',
  },
  compilationError: {
    id: `${scope}.compilationError`,
    defaultMessage: 'Compilation Error',
  },
  runtimeError: {
    id: `${scope}.runtimeError`,
    defaultMessage: 'Runtime Error',
  },
  tryAgain: {
    id: `${scope}.tryAgain`,
    defaultMessage: 'Try Again',
  },
  areYouSure: {
    id: `${scope}.areYouSure`,
    defaultMessage: 'Are you sure to delete this test case?',
  },
  yes: {
    id: `${scope}.yes`,
    defaultMessage: 'Yes',
  },
  no: {
    id: `${scope}.no`,
    defaultMessage: 'No',
  },
})
