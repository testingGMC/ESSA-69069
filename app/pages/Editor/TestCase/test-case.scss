.test-case{
    .test-btn {
        width: 100px;
        margin-top: 12px;
    }
    .answer {
        font-size: 16px;
        font-weight: bold; 
        margin-left: 15px;
    }
    .green-border {
        border: 2px solid green;
    }
    .red-border {
        border: 2px solid red;
    }
    .default-border {
        border: 2px solid #d9d9d9;
    }
}
