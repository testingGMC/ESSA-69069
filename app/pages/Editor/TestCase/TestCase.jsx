/**
 *
 * TestCase
 *
 */

import React, { memo, useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { isEmpty } from 'lodash'
import { FormattedMessage, injectIntl, intlShape } from 'react-intl'
import { Button, Input, Popconfirm } from 'antd'
import Axios from 'axios'
import {
  JAVASCRIPT,
  LANGUAGES,
  CORRECT_ANSWER,
  WRONG_ANSWER,
  COMPILING_ERROR,
  IN_QUEUE,
  PROCESSING,
  INTERNAL_ERROR,
  TIME_LIMIT_EXCEEDED,
} from 'shared/redux/editor/utils/editorConstants'
import codeCompilingmessages from '../CodeCompiling/messages'
import messages from './messages'
import './test-case.scss'

export function TestCase({
  el,
  assessment,
  skill,
  updateCodeCompilingUpdateUseCase,
  updateCodeCompilingDeleteUseCase,
  intl,
}) {
  const [isTest, setIsTest] = useState(false)
  const [token, setToken] = useState('')
  const [responseDescription, setResponseDescription] = useState('')
  const [responseId, setResponseId] = useState(0)
  const [correctResponse, setCorrectResponse] = useState('')

  const messagesList = {
    correctAnswer: intl.formatMessage({
      ...messages.correctAnswer,
    }),
    wrongAnswer: intl.formatMessage({
      ...messages.wrongAnswer,
    }),
    compilationError: intl.formatMessage({
      ...messages.compilationError,
    }),
    runtimeError: intl.formatMessage({
      ...messages.runtimeError,
    }),
    tryAgain: intl.formatMessage({
      ...messages.tryAgain,
    }),
    areYouSure: intl.formatMessage({
      ...messages.areYouSure,
    }),
    yes: intl.formatMessage({
      ...messages.yes,
    }),
    no: intl.formatMessage({
      ...messages.no,
    }),
  }

  function testApi() {
    Axios.get(
      `https://api.judge0.com/submissions/${token}?base64_encoded=false`,
    )
      .then(res => {
        setResponseId(res.data.status.id)
        setCorrectResponse(res.data.stdout)
        setIsTest(false)
        if (
          res.data.status.id === IN_QUEUE ||
          res.data.status.id === PROCESSING ||
          res.data.status.id === TIME_LIMIT_EXCEEDED
        ) {
          setTimeout(() => {
            testApi()
          }, 2000)
        } else if (res.data.status.id === CORRECT_ANSWER) {
          setResponseDescription(messagesList.correctAnswer)
        } else if (res.data.status.id === WRONG_ANSWER) {
          setResponseDescription(messagesList.wrongAnswer)
        } else if (res.data.status.id === COMPILING_ERROR) {
          setResponseDescription(messagesList.compilationError)
        } else if (
          res.data.status.id >= COMPILING_ERROR &&
          res.data.status.id <= INTERNAL_ERROR
        ) {
          setResponseDescription(messagesList.runtimeError)
        } else setResponseDescription(messagesList.tryAgain)
      })
      .catch(() => {
        setResponseDescription(messagesList.tryAgain)
        setIsTest(false)
      })
  }

  useEffect(() => {
    if (isTest) {
      Axios.post('https://api.judge0.com/submissions', {
        source_code: skill.data[assessment.solutionId].text,
        language_id: isEmpty(skill.data[assessment.languageId].text)
          ? JAVASCRIPT
          : LANGUAGES[skill.data[assessment.languageId].text].label,
        stdin: skill.data[el].input,
        expected_output: skill.data[el].output,
      })
        .then(res => {
          setToken(res.data.token)
        })
        .catch(() => {
          setResponseDescription(messagesList.tryAgain)
          setIsTest(false)
        })
    }
  }, [isTest])

  useEffect(() => {
    if (token !== '') {
      setTimeout(() => {
        testApi()
      }, 3000)
    }
  }, [token])

  return (
    <div className="test-case input-output-row">
      <Input.TextArea
        type="text"
        className="input"
        value={skill.data[el] ? skill.data[el].input : ''}
        onChange={e =>
          updateCodeCompilingUpdateUseCase(el, assessment.id, {
            input: e.target.value,
            output: skill.data[el].output,
          })
        }
      />
      <Input.TextArea
        type="text"
        className={`output ${
          responseId === CORRECT_ANSWER
            ? 'green-border'
            : responseId === WRONG_ANSWER
              ? 'red-border'
              : 'default-border'
        }`}
        value={skill.data[el] ? skill.data[el].output : ''}
        onChange={e =>
          updateCodeCompilingUpdateUseCase(el, assessment.id, {
            input: skill.data[el].input,
            output: e.target.value,
          })
        }
      />
      <div>
        <Button
          loading={isTest}
          className="test-btn"
          onClick={() => {
            setIsTest(true)
          }}
          type="primary"
        >
          <FormattedMessage {...messages.test} />
        </Button>
        <span className="answer">{responseDescription}</span>
        {responseId === 4 && (
          <span className="answer">
            ({messagesList.correctAnswer} : ({correctResponse}) )
          </span>
        )}
      </div>
      {assessment.testCases && assessment.testCases.length !== 1 && (
        <Popconfirm
          title={messagesList.areYouSure}
          onConfirm={() => {
            updateCodeCompilingDeleteUseCase(skill.data[el].id, assessment.id)
          }}
          okText={messagesList.yes}
          cancelText={messagesList.no}
        >
          <Button className="delete-btn" type="danger">
            <FormattedMessage {...codeCompilingmessages.delete} />
          </Button>
        </Popconfirm>
      )}
    </div>
  )
}

TestCase.propTypes = {
  el: PropTypes.string.isRequired,
  assessment: PropTypes.object.isRequired,
  skill: PropTypes.object.isRequired,
  updateCodeCompilingUpdateUseCase: PropTypes.func.isRequired,
  updateCodeCompilingDeleteUseCase: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
}

export default memo(injectIntl(TestCase))
