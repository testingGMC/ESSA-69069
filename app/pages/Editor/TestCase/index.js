import TestCase from './TestCase'

export default TestCase
