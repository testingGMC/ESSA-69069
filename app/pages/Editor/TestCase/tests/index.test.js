import React from 'react'
import { render } from 'react-testing-library'
import { IntlProvider } from 'react-intl'
import TestCase from '../index'
import { DEFAULT_LOCALE } from '../../../../i18n'

const initialProps = {
  el: '',
  assessment: {},
  updateCodeCompilingUpdateUseCase: jest.fn(),
  updateCodeCompilingDeleteUseCase: jest.fn(),
  skill: { data: { hello: { input: 5, output: 8 } } },
  intl: 0,
}

describe('<TestCase />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error')
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <TestCase {...initialProps} />
      </IntlProvider>,
    )
    expect(spy).not.toHaveBeenCalled()
  })
  it('Expect to have additional unit tests specified', () => {
    expect(true).toEqual(true)
  })

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <TestCase {...initialProps} />
      </IntlProvider>,
    )
    expect(firstChild).toMatchSnapshot()
  })
})
