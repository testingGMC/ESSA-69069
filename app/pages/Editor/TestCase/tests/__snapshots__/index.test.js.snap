// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TestCase /> Should render and match the snapshot 1`] = `
<div
  class="test-case input-output-row"
>
  <textarea
    class="ant-input input"
    type="text"
  />
  <textarea
    class="ant-input output default-border"
    type="text"
  />
  <div>
    <button
      class="ant-btn test-btn ant-btn-primary"
      type="button"
    >
      <span>
        Test
      </span>
    </button>
    <span
      class="answer"
    />
  </div>
</div>
`;
